src/runtime-rs/crates/hypervisor/src/qemu/qmp.rs

```diff
@@ -128,6 +142,153 @@ impl Qmp {
+                    dimm_info.data.hotplugged
+                );
+
+                if dimm_info.data.hotpluggable && dimm_info.data.hotplugged {
+                    hotplugged_mem_size += dimm_info.data.size as u64;
+                }
+            }
+        }
+
+        Ok(hotplugged_mem_size)
+    }
+
+    pub fn hotplug_memory(&mut self, size: u64) -> Result<()> {
+        let memdev_idx = self
+            .qmp
+            .execute(&qapi_qmp::query_memory_devices {})?
+            .into_iter()
+            .filter(|memdev| {
+                if let qapi_qmp::MemoryDeviceInfo::dimm(dimm_info) = memdev {
+                    return dimm_info.data.hotpluggable && dimm_info.data.hotplugged;
+                }
+                false
+            })
+            .count();
+
+        let memory_backend_id = format!("hotplugged-{}", memdev_idx);
+
+        let memory_backend = qmp::object_add(qapi_qmp::ObjectOptions::memory_backend_file {
+            id: memory_backend_id.clone(),
+            memory_backend_file: qapi_qmp::MemoryBackendFileProperties {
+                base: qapi_qmp::MemoryBackendProperties {
+                    dump: None,
+                    host_nodes: None,
+                    merge: None,
+                    policy: None,
+                    prealloc: None,
+                    prealloc_context: None,
+                    prealloc_threads: None,
+                    reserve: None,
+                    share: Some(true),
+                    x_use_canonical_path_for_ramblock_id: None,
+                    size,
+                },
+                align: None,
+                discard_data: None,
+                offset: None,
+                pmem: None,
+                readonly: None,
+                mem_path: "/dev/shm".to_owned(),
+            },
+        });
+        self.qmp.execute(&memory_backend)?;
+
+        let memory_frontend_id = format!("frontend-to-{}", memory_backend_id);
+
+        let mut mem_frontend_args = Dictionary::new();
+        mem_frontend_args.insert("memdev".to_owned(), memory_backend_id.into());
+        self.qmp.execute(&qmp::device_add {
+            bus: None,
+            id: Some(memory_frontend_id),
+            driver: "pc-dimm".to_owned(),
+            arguments: mem_frontend_args,
+        })?;
+
+        Ok(())
+    }
+
+    pub fn hotunplug_memory(&mut self, size: i64) -> Result<()> {
+        let frontend = self
+            .qmp
+            .execute(&qapi_qmp::query_memory_devices {})?
+            .into_iter()
+            .find(|memdev| {
+                if let qapi_qmp::MemoryDeviceInfo::dimm(dimm_info) = memdev {
+                    let dimm_id = match dimm_info.data.id {
+                        Some(ref id) => id,
+                        None => return false,
+                    };
+                    if dimm_info.data.hotpluggable
+                        && dimm_info.data.hotplugged
+                        && dimm_info.data.size == size
```

Comment:
Just out of curiosity: should the size of memory requested for unplug be the same as that of memory for the plugged device?

Reply:
I'm not overly knowledgeable about the fine points of this area, however it appears obvious from the nature of hotunplugging implementation that it's impossible to hotunplug an arbitrary amount of memory.  So the sizes need to match and it appears to be a user responsibility to ensure that that's the case.

This is fulfilled in runtime-rs because runtime-rs hotplugs memory requests of individual containers and then hotunplugs them again so match is guaranteed.  E.g. if a container in a pod requests 1G and another containers in the same pod requests 2G, runtime-rs will hotplug 1G and then 2G on launch and then hotunplug 1G and then 2G on pod termination.

The matching by size works in practice, a memory block to hotunplug is always found on the first hotunplug attempt.  The fact that hotunplugging it usually fails is more of a problem.  If that happens subsequent hotunplugging requests don't even find a matching-size block anymore.

Reply:
Oh, it makes sense. The pod example definitely boosts my understanding. Thanks!